File: Arrays_2/coin_flip_streaks.py
import random, time

start_time = time.time()

numberOfStreaks = 0
for experimentNumber in range(10000):
    # Code that creates a list of 100 'heads' or 'tails' values.
    flips = []
    for i in range(100):
        if random.randint(0, 1):
            flips.append("H")
        else:
            flips.append("T")

    # Code that checks if there is a streak of 6 heads or tails in a row.
    for i in range(100 - 6):
        if (
            flips[i]
            == flips[i + 1]
            == flips[i + 2]
            == flips[i + 3]
            == flips[i + 4]
            == flips[i + 5]
        ):
            numberOfStreaks += 1
            break

end_time = time.time()

print("Chance of streak: %s%%" % (numberOfStreaks / 100))

print("Elapsed time: %s seconds" % (end_time - start_time))

File: Arrays_2/src/bin/coin_flip_streaks.rs
use rand::Rng;
use std::time::Instant;

fn main() {
    let start_time = Instant::now();
    let mut numberOfStreaks = 0;
    for _ in 0..10000 {
        // Code that creates a list of 100 'heads' or 'tails' values.
        let mut flips = Vec::new();
        for _ in 0..100 {
            let flip = if rand::thread_rng().gen_bool(0.5) {
                "H"
            } else {
                "T"
            };
            flips.push(flip);
        }
        // Code that checks if there is a streak of 6 heads or tails in a row.
        for i in 0..flips.len() - 6 {
            if flips[i..i + 6].iter().all(|&x| x == "H")
                || flips[i..i + 6].iter().all(|&x| x == "T")
            {
                numberOfStreaks += 1;
                break;
            }
        }
    }
    let elapsed_time = start_time.elapsed();
    println!("Chance of streak: {}%", numberOfStreaks / 100);
    println!("Time taken: {:.6} seconds", elapsed_time.as_secs_f64());
}
